hathor/cli/side_dag.py

```diff
@@ -0,0 +1,220 @@
+# Copyright 2024 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import signal
+import sys
+import traceback
+from dataclasses import dataclass
+from multiprocessing import Pipe, Process
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    # Workaround for a typing issue in Windows
+    if sys.platform == 'win32':
+        from multiprocessing.connection import _ConnectionBase as Connection
+    else:
+        from multiprocessing.connection import Connection
+
+import psutil
+from structlog import get_logger
+
+from hathor.cli.run_node import RunNode
+
+logger = get_logger()
+
+PRE_SETUP_LOGGING: bool = False
+HATHOR_NODE_INIT_TIMEOUT: int = 10
+
+
+@dataclass(frozen=True, slots=True)
+class HathorProcessInitFail:
+    reason: str
+
+
+@dataclass(frozen=True, slots=True)
+class HathorProcessInitSuccess:
+    pass
+
+
+@dataclass(frozen=True, slots=True)
+class HathorProcessTerminated:
+    pass
+
+
+@dataclass(frozen=True, slots=True)
+class SideDagProcessTerminated:
+    pass
+
+
+class SideDagRunNode(RunNode):
+    env_vars_prefix = 'hathor_side_dag_'
+
+
+def main(capture_stdout: bool) -> None:
+    """
+    This command runs two full node instances in separate processes.
+
+    The main process runs a side-dag full node, and it accepts the same options as the `run_node` command. Options
+    with the `--side-dag` prefix are passed to the side-dag full node, while options without this prefix are passed
+    to the non-side-dag full node, which runs in a background process and is commonly just a Hathor full node.
+    Whenever one of the full nodes fail, the other is automatically terminated.
+
+    By default, both full nodes output logs to stdout, but logs can be configured independently. Here's an example:
+
+    ```bash
+    $ python -m hathor side_dag
+        --testnet
+        --procname-prefix testnet-
+        --memory-storage
+        --disable-logs
+        --side-dag-config-yaml ./my-side-dag.yml
+        --side-dag-procname-prefix my-side-dag-
+        --side-dag-memory-storage
+        --side-dag-json-logs
```

Comment:
Does `--side-dag-json-logs` need to be a separate option from `--json-logs`? The reason `--json-logs` exist is to make all the logs that the stdout/stderr (as observed by whoever started the process) be written as `JSON` entries (one per line), instead of formatted entries (which can be colored). The option exists independently of which subcommand is being used even for those not based on `RunNode`.

What do you think about using only one argument instead?

Reply:
It could be a single option, the reason I implemented it separately is to be able to configure each full node's logging independently. I created the `LoggingOutput` enum and those arguments simply select one of the enum options. It accepts either `--[side-dag-]disable-logs`, `--[side-dag-]json-logs`, or not set, which respectively map to `NULL`, `JSON`, or `PRETTY` (the default).

So for example, if I'm running locally, I would probably set only `--disable-logs`, which would disable any logs from the hathor node, and the side-dag logs would be pretty outputted to stdout/sderr.

If I'm running on a server, I would probably set `--json-logs --side-dag-json-logs`, to get json outputs from both nodes. But one could easily want to discard hathor logs, for example, in which case they would set `--disable-logs --side-dag-json-logs`.

In summary, this is just giving more flexibility. Do you prefer to keep a single option?

Reply:
I see, I'm not against the flexibility, but I'm not a fan of having the behavior of `--json-logs` sort of change (from having all logs in JSON, to some of the logs in JSON), I understand that production would just need to use `--json-logs --side-dag-json-logs` but I think that's a bit counter-intuitive.

I'd rather have the `side-dag` process use the same behavior (`NULL`, `JSON`, `PRETTY`) as the main log, unless there's an argument that changes it.

What about making the options `--json-logs` and `--disable-logs` take an optional argument: `both`, `main`, `side-dag`? And the default when the argument is present would be `both`. Since these two are special options that are processed before the `RunNode` ones, I think it would be an acceptable compromise.

Reply:
This can be implemented with [ArgumentParser.parse_known_args](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.parse_known_args) instead of doing it manually like is was before.

Reply:
Done in https://github.com/HathorNetwork/hathor-core/pull/1039/commits/90a0606cb53cc1939157c4b172cf5ad8d380e6f0